response.go

```diff
@@ -215,253 +215,378 @@ func visitModelNode(model interface{}, included *map[string]*Node,
+	} else {
+		kind = fieldType.Type.Kind()
+	}
+
+	// Handle allowed types
+	switch kind {
+	case reflect.String:
+		node.ID = v.Interface().(string)
+	case reflect.Int:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int)), 10)
+	case reflect.Int8:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int8)), 10)
+	case reflect.Int16:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int16)), 10)
+	case reflect.Int32:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int32)), 10)
+	case reflect.Int64:
+		node.ID = strconv.FormatInt(v.Interface().(int64), 10)
+	case reflect.Uint:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint)), 10)
+	case reflect.Uint8:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint8)), 10)
+	case reflect.Uint16:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint16)), 10)
+	case reflect.Uint32:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint32)), 10)
+	case reflect.Uint64:
+		node.ID = strconv.FormatUint(v.Interface().(uint64), 10)
+	case reflect.Struct:
+		if nStr, ok := v.Interface().(sql.NullString); ok {
+			node.ID = nStr.String
+			break
+		}
+
+		if nI32, ok := v.Interface().(sql.NullInt32); ok {
+			node.ID = strconv.FormatInt(int64(nI32.Int32), 10)
+			break
+		}
+
+		if nI64, ok := v.Interface().(sql.NullInt64); ok {
+			node.ID = strconv.FormatInt(nI64.Int64, 10)
 			break
 		}
 
-		if annotation == annotationPrimary {
-			v := fieldValue
+		fallthrough
+	default:
+		// We had a JSON float (numeric), but our field was not one of the
+		// allowed numeric types
+		return nil, ErrBadJSONAPIID
+	}
+
+	return node, nil
+}
+
+func resolveNodeAttribute(node *Node, fieldValue reflect.Value, args []string) *Node {
+	var omitEmpty, iso8601 bool
+
+	if len(args) > 2 {
+		for _, arg := range args[2:] {
+			switch arg {
+			case annotationOmitEmpty:
+				omitEmpty = true
+			case annotationISO8601:
+				iso8601 = true
+			}
+		}
+	}
+
+	if node.Attributes == nil {
+		node.Attributes = make(map[string]interface{})
+	}
+
+	switch fieldValue.Type() {
+	case reflect.TypeOf(time.Time{}):
+		t := fieldValue.Interface().(time.Time)
+
+		if t.IsZero() {
+			return node
+		}
+
+		if iso8601 {
+			node.Attributes[args[1]] = t.UTC().Format(iso8601TimeFormat)
+		} else {
+			node.Attributes[args[1]] = t.Unix()
+		}
+	case reflect.TypeOf(new(time.Time)):
+		// A time pointer may be nil
+		if fieldValue.IsNil() {
+			if omitEmpty {
+				return node
+			}
+
+			node.Attributes[args[1]] = nil
+		} else {
+			t := fieldValue.Interface().(*time.Time)
+
+			if t.IsZero() && omitEmpty {
+				return node
+			}
 
-			// Deal with PTRS
-			var kind reflect.Kind
-			if fieldValue.Kind() == reflect.Ptr {
-				kind = fieldType.Type.Elem().Kind()
-				v = reflect.Indirect(fieldValue)
+			if iso8601 {
+				node.Attributes[args[1]] = t.UTC().Format(iso8601TimeFormat)
 			} else {
-				kind = fieldType.Type.Kind()
+				node.Attributes[args[1]] = t.Unix()
+			}
+		}
+	case reflect.TypeOf(sql.NullTime{}):
```

Comment:
I'm not sure how common the usage of `*sql.NullTime` is (if that even makes sense), so for now I just added support for `sql.NullTime`.